plugins/woo-ai/src/shared/productData.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Internal dependencies
+ */
+import { Attribute, ProductData, TinyContent } from './types';
+
+declare const tinymce: { get: ( str: string ) => TinyContent };
+
+const isElementVisible = ( element: HTMLElement ) =>
+	! ( window.getComputedStyle( element ).display === 'none' );
+
+const getCategories = (): string[] => {
+	const categoryCheckboxEls: NodeListOf< HTMLInputElement > =
+		document.querySelectorAll(
+			'#taxonomy-product_cat input[name="tax_input[product_cat][]"]:checked'
+		);
+
+	const tempCategories: string[] = [];
+
+	categoryCheckboxEls.forEach( ( el ) => {
+		if ( ! el.value.length ) {
+			return;
+		}
+
+		tempCategories.push( el.value );
+	} );
+
+	return tempCategories;
+};
+
+const getTags = (): string[] => {
+	const tagsEl: HTMLTextAreaElement | null = document.querySelector(
+		'textarea[name="tax_input[product_tag]"]'
+	);
+
+	const tags = tagsEl ? tagsEl.value.split( ',' ) : [];
+
+	return tags.filter( ( tag ) => tag !== '' );
+};
+
+const getAttributes = (): Attribute[] => {
+	const attributeSelectEls: NodeListOf< HTMLSelectElement > =
+		document.querySelectorAll(
+			"#product_attributes select[name^='attribute_values']"
+		);
+
+	const tempAttributes: Attribute[] = [];
+
+	attributeSelectEls.forEach( ( el: HTMLSelectElement ) => {
+		const attributeName =
+			el.getAttribute( 'data-taxonomy' )?.replace( 'pa_', '' ) || '';
+
+		const attributeValues = Array.from( el.selectedOptions )
+			.map( ( option ) => option.text )
+			.join( ',' );
+
+		if ( ! attributeValues || ! attributeName ) {
+			return;
+		}
+
+		tempAttributes.push( {
+			name: attributeName,
+			value: attributeValues,
+		} );
+	} );
+
+	return tempAttributes;
+};
+
+const getDescription = (): string => {
+	const isBlockEditor =
+		document.querySelectorAll( '.block-editor' ).length > 0;
+
+	let description_value = '';
+
+	if ( ! isBlockEditor ) {
+		const content = document.querySelector(
+			'#content'
+		) as HTMLInputElement;
+		if ( content && isElementVisible( content ) ) {
+			description_value = content.value;
+		} else if ( typeof tinymce === 'object' && tinymce.get( 'content' ) ) {
```

Comment:
I have a `useTinyEditor` hook available in the `hooks` folder to make this easier.

Reply:
I don't [think](https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-from-react-functions) I can use that here since it's a normal function. Maybe we can change the TinyEditor hook to use a function from the productData function to get the tinyEditor content.

Reply:
Ah! Understood. We could just refactor to keep them in a utils file, and then base the hook off of that.  Is that what you mean? We can look at that as a follow-up.

Reply:
Yes exactly. Let's do that in a follow-up.